### PR TITLE
Libimagstore/hook errors not aborting

### LINKED ISSUE #368: Hook system: Hooks should be able to return errors without aborting the store operation
For example if my Hook A fails, the fail isn't critical but should nevertheless be reported to the user, the current way of doing this is by calling `warn!()` or `info!()` within the hook.

We should refactor this, so a Hook can return an `Err()` without causing the store operation to abort.

I guess we should refactor the error type of the hook operation to include an `abort` field which can be set to a boolean value and causes the operation to be aborted if it is set to true, otherwise the error is simply printed to `stderr`.

This is a complex change which requires a big refactoring in the ´libimagstore` and `libimagstdhook`. Maybe we can get this working without rewriting parts of `libimagstdhook`, but I'm not sure about this.

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Closes #368

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Backwards compatible. Yay! :smile:

**Comment by matthiasbeyer** [comment]
@TheNeikos would be nice if you could have a look whether I've overlooked something.

File: libimagstore/src/hook/error.rs [hunk 1]
```diff
@@ -80,6 +82,14 @@ impl HookError {
         self.err_type.clone()
     }
 
+    pub fn aborting(&mut self, b: bool) {
```
Comment (TheNeikos) [inline libimagstore/src/hook/error.rs]: Uhm, why should I be able to call `aborting(false)` ?

Comment (matthiasbeyer) [inline libimagstore/src/hook/error.rs]: To set the error to be non-aborting. The default is that a `HookError` aborts the store action (create, update, whatever). If you set the error to non-aborting, the error gets `trace()`ed, but doesn't about the store action.

**Comment by matthiasbeyer** [comment]
This is _not_ ready as the `Aspect` code has to be adapted to go on calling hooks even if there was an error (as long as the error is non-aborting).

**Comment by matthiasbeyer** [comment]
Waiting on #414

**Comment by matthiasbeyer** [comment]
Okay, we are ready here. @TheNeikos review pls!

File: libimagstore/src/store.rs [hunk 2]
```diff
@@ -299,7 +302,13 @@ impl Store {
     pub fn create<'a, S: IntoStoreId>(&'a self, id: S) -> Result<FileLockEntry<'a>> {
         let id = self.storify_id(id.into_storeid());
         if let Err(e) = self.execute_hooks_for_id(self.pre_create_aspects.clone(), &id) {
-            return Err(e);
+            if e.is_aborting() {
+                return Err(e).map_err(|e| {
+                    SE::new(SEK::PreHookExecuteError, Some(Box::new(e)))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: We can shorten this with the new error setup: 

```
.map_err(Box::new)
.map_err(|e| SEK::ErrorKind.into_error_with_cause(e))
```

File: libimagstore/src/store.rs [hunk 3]
```diff
@@ -330,7 +339,13 @@ impl Store {
     pub fn retrieve<'a, S: IntoStoreId>(&'a self, id: S) -> Result<FileLockEntry<'a>> {
         let id = self.storify_id(id.into_storeid());
         if let Err(e) = self.execute_hooks_for_id(self.pre_retrieve_aspects.clone(), &id) {
-            return Err(e);
+            if e.is_aborting() {
+                return Err(e).map_err(|e| {
+                    SE::new(SEK::PreHookExecuteError, Some(Box::new(e)))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Shorten this, see above.

File: libimagstore/src/store.rs [hunk 4]
```diff
@@ -357,7 +372,13 @@ impl Store {
     {
         let id = self.storify_id(id.into_storeid());
         if let Err(e) = self.execute_hooks_for_id(self.pre_retrieve_aspects.clone(), &id) {
-            return Err(e);
+            if e.is_aborting() {
+                return Err(e).map_err(|e| {
+                    SE::new(SEK::PreHookExecuteError, Some(Box::new(e)))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Shorten this, see above.

File: libimagstore/src/store.rs [hunk 5]
```diff
@@ -450,14 +471,21 @@ impl Store {
     /// Return the `FileLockEntry` and write to disk
     pub fn update<'a>(&'a self, mut entry: FileLockEntry<'a>) -> Result<()> {
         if let Err(e) = self.execute_hooks_for_mut_file(self.pre_update_aspects.clone(), &mut entry) {
-            return Err(e);
+            if e.is_aborting() {
+                return Err(e).map_err(|e| {
+                    SE::new(SEK::PreHookExecuteError, Some(Box::new(e)))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Shorten this, see above.

File: libimagstore/src/store.rs [hunk 6]
```diff
@@ -590,29 +625,29 @@ impl Store {
...
     {
         match aspects.lock() {
-            Err(_) => return Err(SE::new(SEK::PreHookExecuteError, None)),
+            Err(_) => return Err(HookErrorKind::HookExecutionError.into()),
             Ok(g) => g
         }.iter().fold(Ok(()), |acc, aspect| {
             debug!("[Aspect][exec]: {:?}", aspect);
             acc.and_then(|_| (aspect as &StoreIdAccessor).access(id))
-        }).map_err(|e| SE::new(SEK::PreHookExecuteError, Some(Box::new(e))))
+        }).map_err(|e| HookError::new(HookErrorKind::HookExecutionError, Some(Box::new(e))))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Shorten this, see above.

File: libimagstore/src/store.rs [hunk 7]
```diff
@@ -590,29 +625,29 @@ impl Store {
...
     {
         match aspects.lock() {
-            Err(_) => return Err(SE::new(SEK::PreHookExecuteError, None)),
+            Err(_) => return Err(HookErrorKind::HookExecutionError.into()),
             Ok(g) => g
         }.iter().fold(Ok(()), |acc, aspect| {
             debug!("[Aspect][exec]: {:?}", aspect);
             acc.and_then(|_| aspect.access_mut(fle))
-        }).map_err(|e| SE::new(SEK::PreHookExecuteError, Some(Box::new(e))))
+        }).map_err(|e| HookError::new(HookErrorKind::HookExecutionError, Some(Box::new(e))))
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Shorten this, see above.

**Comment by TheNeikos** [comment]
What about a macro that works similarly to `try!`? It could return on abort, and otherwise return a `Result<T>`

**Pull request merged**